src/MyArray.js

```diff
@@ -1,112 +1,245 @@
+  MyArray.prototype.find = function (fn) {
+    return this.get(this.findIndex(fn));
+  };
 
-};
+  MyArray.prototype.findIndex = function (fn) {
+    const firstIndex = allIndicesWhere(fn, this).get(0);
+    if (firstIndex === undefined) {
+      return -1;
+    }
+    return firstIndex;
+  };
 
-MyArray.prototype.unshift = function (element) {
+  MyArray.prototype.equals = function (other) {
+    const matches = allIndicesWhere(isEqualToI(this), other);
+    return matches.length() === this.length() && matches.length() === other.length();
+  };
 
-};
+  MyArray.prototype.forEach = function (fn) {
+    this.map(fn);
+  };
 
-MyArray.prototype.slice = function (start, end) {
+  MyArray.prototype.join = function (separator) {
+    if (separator === undefined) {
+      separator = ",";
+    }
+    const length = this.length();
+    var strVal = "";
+    this.forEach(function (item, i) {
+      strVal += item;
+      if (i < length - 1) {
+        strVal += separator;
+      }
+    });
+    return strVal;
+  };
+
+  MyArray.prototype.toString = function () {
+    return this.join();
+  };
+
+  MyArray.prototype.map = function (fn) {
+    const newArray = new MyArray(this.length());
+    for (var i = 0, l = this.length(); i < l; i++) {
+      newArray.set(i, fn(this.get(i), i));
+    }
+    return newArray;
+  };
+
+  MyArray.prototype.filter = function (fn) {
+    const indices = allIndicesWhere(fn, this);
+    const get = this.get.bind(this);
+    return indices.map(get);
+  };
+
+  MyArray.prototype.some = function (fn) {
+    return allIndicesWhere(fn, this).length() > 0;
+  };
+
+  MyArray.prototype.every = function (fn) {
+    return allIndicesWhere(fn, this).length() === this.length();
+  };
+
+  MyArray.prototype.fill = function (fillValue, start, end) {
+    if (start === undefined) {
+      start = 0;
+    }
+    if (end === undefined) {
+      end = this.length();
+    }
+    this.elements = this.map(function (item, i) {
+      if (i < start || i >= end) {
+        return item;
+      }
+      return fillValue;
+    }).elements;
+  };
+
+  MyArray.prototype.reverse = function () {
+    const length = this.length();
+    const get = this.get.bind(this);
+    this.elements = this.map(function (item, i) {
+      return get(length - 1 - i);
+    }).elements;
+  };
+
+  MyArray.prototype.shift = function () {
+    const firstEle = this.get(0);
+    this.elements = this.filter(function (item, i) {
+      return i > 0;
+    }).elements;
+    return firstEle;
+  };
+
+  MyArray.prototype.unshift = function (element) {
+    this.elements = MyArray.of(element).concat(this).elements;
+  };
+
+  MyArray.prototype.slice = function (start, end) {
+    if (start === undefined) {
+      start = -1;
+    }
+    if (end === undefined) {
+      end = this.length();
+    }
+    return this.filter(function (item, i) {
+      return i >= start && i < end;
+    });
+  };
+
+  MyArray.prototype.splice = function (startIndex, deleteCount) {
```

Comment:
This is a GOLD way to implement `splice`. Love it! Short, concise, and simple.

I would probably just make the `middle` part as a `for loop + push`, so that uninitiated in `apply` can understand it, as well :)